File: py/deepgene/gene_data.py
# ABOUTME: Module for fetching gene data from MyGene.info API
# ABOUTME: Provides deterministic gene information for use as context in AI queries

import logging
from pydantic import BaseModel
from typing import Literal

logger = logging.getLogger(__name__)


class GeneData(BaseModel):
    """Structured gene data from MyGene.info database"""
    gene_symbol: str
    gene_name: str | None = None
    summary: str | None = None
    go_terms: dict[str, list] | None = None
    pathways: list[str] | None = None
    mim_diseases: list[str] | None = None
    generif: list[str] | None = None
    entrezgene_id: int | None = None
    ensembl_id: str | None = None
    genomic_location: str | None = None
    source: Literal["mygene.info"] = "mygene.info"


MYGENE_FIELDS = (
    'symbol,name,summary,'
    'go.BP,go.MF,go.CC,'
    'pathway.reactome,pathway.wikipathways,pathway.kegg,'
    'MIM,generif,genomic_pos,'
    'entrezgene,ensembl.gene'
)


def fetch_gene_data(gene_symbol: str) -> GeneData | None:
    """
    Fetch gene data from MyGene.info API.

    Args:
        gene_symbol: Gene symbol to query (e.g., "CTNND2", "BRCA1")

    Returns:
        GeneData object if successful, None on error or not found

    Fields requested:
        - symbol, name, summary
        - go (BP, MF, CC)
        - pathway (reactome, wikipathways, kegg)
        - MIM (OMIM disease associations)
        - generif (research summaries)
        - genomic_pos, entrezgene, ensembl.gene
    """
    try:
        import mygene
        mg = mygene.MyGeneInfo()

        query_result = mg.query(
            f'symbol:{gene_symbol}',
            species='human',
            fields='entrezgene',
            size=1
        )

        if not query_result or 'hits' not in query_result or not query_result['hits']:
            logger.warning(f"No data found for gene symbol: {gene_symbol}")
            return None

        gene_id = query_result['hits'][0].get('entrezgene')
        if not gene_id:
            logger.warning(f"No gene ID found for symbol: {gene_symbol}")
            return None

        result = mg.getgene(
            gene_id,
            fields=MYGENE_FIELDS,
            timeout=5
        )

        if not result:
            logger.warning(f"No detailed data found for gene ID: {gene_id}")
            return None

        return parse_mygene_response(result, gene_symbol)

    except TimeoutError:
        logger.warning(f"Timeout fetching data for {gene_symbol}")
        return None
    except Exception as e:
        logger.error(f"Error fetching gene data for {gene_symbol}: {e}")
        return None


def parse_mygene_response(result: dict, gene_symbol: str) -> GeneData:
    """
    Parse MyGene.info API response into GeneData model.

    Args:
        result: Raw response from MyGene.info
        gene_symbol: Original query symbol

    Returns:
        GeneData object with parsed fields
    """
    go_terms = {}
    if 'go' in result:
        go_data = result['go']
        if 'BP' in go_data:
            go_terms['BP'] = [term.get('term', '') for term in go_data['BP'][:10]]
        if 'MF' in go_data:
            go_terms['MF'] = [term.get('term', '') for term in go_data['MF'][:10]]
        if 'CC' in go_data:
            go_terms['CC'] = [term.get('term', '') for term in go_data['CC'][:10]]

    pathways = []
    if 'pathway' in result:
        pathway_data = result['pathway']
        if isinstance(pathway_data, dict):
            for source in ['reactome', 'wikipathways', 'kegg']:
                if source in pathway_data:
                    source_pathways = pathway_data[source]
                    if isinstance(source_pathways, list):
                        for pw in source_pathways:
                            if isinstance(pw, dict) and 'name' in pw:
                                pathways.append(f"{pw['name']} ({source.capitalize()})")
                    elif isinstance(source_pathways, dict) and 'name' in source_pathways:
                        pathways.append(f"{source_pathways['name']} ({source.capitalize()})")

    mim_diseases = []
    if 'MIM' in result:
        mim_data = result['MIM']
        if isinstance(mim_data, list):
            for mim in mim_data[:5]:
                if isinstance(mim, dict):
                    mim_id = mim.get('MIM', '')
                    mim_name = mim.get('name', '')
                    if mim_id or mim_name:
                        mim_diseases.append(f"MIM:{mim_id} - {mim_name}" if mim_id else mim_name)
        elif isinstance(mim_data, dict):
            mim_id = mim_data.get('MIM', '')
            mim_name = mim_data.get('name', '')
            if mim_id or mim_name:
                mim_diseases.append(f"MIM:{mim_id} - {mim_name}" if mim_id else mim_name)

    generif = []
    if 'generif' in result:
        generif_data = result['generif']
        if isinstance(generif_data, list):
            for rif in generif_data[:5]:
                if isinstance(rif, dict):
                    pmid = rif.get('pubmed', '')
                    text = rif.get('text', '')
                    if pmid and text:
                        generif.append(f"PMID:{pmid}: {text[:100]}...")

    ensembl_id = None
    if 'ensembl' in result:
        ensembl_data = result['ensembl']
        if isinstance(ensembl_data, dict):
            ensembl_id = ensembl_data.get('gene')
        elif isinstance(ensembl_data, list) and ensembl_data:
            ensembl_id = ensembl_data[0].get('gene') if isinstance(ensembl_data[0], dict) else None

    genomic_location = None
    if 'genomic_pos' in result:
        gpos = result['genomic_pos']
        if isinstance(gpos, dict):
            chr_num = gpos.get('chr', '')
            start = gpos.get('start', '')
            end = gpos.get('end', '')
            if chr_num and start and end:
                genomic_location = f"chr{chr_num}:{start:,}-{end:,}"
        elif isinstance(gpos, list) and gpos:
            gpos = gpos[0]
            chr_num = gpos.get('chr', '')
            start = gpos.get('start', '')
            end = gpos.get('end', '')
            if chr_num and start and end:
                genomic_location = f"chr{chr_num}:{start:,}-{end:,}"

    return GeneData(
        gene_symbol=result.get('symbol', gene_symbol),
        gene_name=result.get('name'),
        summary=result.get('summary'),
        go_terms=go_terms if go_terms else None,
        pathways=pathways if pathways else None,
        mim_diseases=mim_diseases if mim_diseases else None,
        generif=generif if generif else None,
        entrezgene_id=result.get('entrezgene'),
        ensembl_id=ensembl_id,
        genomic_location=genomic_location,
    )


def format_gene_data_for_llm(gene_data: GeneData) -> str:
    """
    Format GeneData into a structured string for LLM context.

    Args:
        gene_data: Parsed gene data from MyGene.info

    Returns:
        Formatted string with gene information for use as LLM context
    """
    lines = ["GENE DATABASE INFORMATION (MyGene.info):", ""]

    lines.append(f"Gene: {gene_data.gene_symbol}")
    if gene_data.gene_name:
        lines.append(f"Full Name: {gene_data.gene_name}")

    if gene_data.entrezgene_id:
        lines.append(f"NCBI Gene ID: {gene_data.entrezgene_id}")
    if gene_data.ensembl_id:
        lines.append(f"Ensembl ID: {gene_data.ensembl_id}")
    if gene_data.genomic_location:
        lines.append(f"Genomic Location: {gene_data.genomic_location}")

    lines.append("")

    if gene_data.summary:
        lines.append("Summary:")
        lines.append(f"  {gene_data.summary}")
        lines.append("")

    if gene_data.go_terms:
        if 'BP' in gene_data.go_terms and gene_data.go_terms['BP']:
            lines.append("Biological Processes (Gene Ontology):")
            for term in gene_data.go_terms['BP'][:5]:
                lines.append(f"  - {term}")
            lines.append("")

        if 'MF' in gene_data.go_terms and gene_data.go_terms['MF']:
            lines.append("Molecular Functions (Gene Ontology):")
            for term in gene_data.go_terms['MF'][:5]:
                lines.append(f"  - {term}")
            lines.append("")

        if 'CC' in gene_data.go_terms and gene_data.go_terms['CC']:
            lines.append("Cellular Components (Gene Ontology):")
            for term in gene_data.go_terms['CC'][:5]:
                lines.append(f"  - {term}")
            lines.append("")

    if gene_data.pathways:
        lines.append("Known Pathways:")
        for pathway in gene_data.pathways[:8]:
            lines.append(f"  - {pathway}")
        lines.append("")

    if gene_data.mim_diseases:
        lines.append("Disease Associations (OMIM):")
        for disease in gene_data.mim_diseases:
            lines.append(f"  - {disease}")
        lines.append("")

    if gene_data.generif:
        lines.append("Recent Research (GeneRIF):")
        for rif in gene_data.generif[:3]:
            lines.append(f"  - {rif}")
        lines.append("")

    return "\n".join(lines)

File: js/src/lib/gene_data.ts
// ABOUTME: Module for fetching gene data from MyGene.info API
// ABOUTME: Provides deterministic gene information for use as context in AI queries

import type { GeneData } from '../types';

const MYGENE_BASE_URL = 'https://mygene.info/v3';
const MYGENE_FIELDS =
  'symbol,name,summary,' +
  'go.BP,go.MF,go.CC,' +
  'pathway.reactome,pathway.wikipathways,pathway.kegg,' +
  'MIM,generif,genomic_pos,' +
  'entrezgene,ensembl.gene';

interface MyGeneQueryResponse {
  hits: Array<{
    entrezgene?: number;
    [key: string]: unknown;
  }>;
}

interface MyGeneDetailResponse {
  symbol?: string;
  name?: string;
  summary?: string;
  entrezgene?: number;
  go?: {
    BP?: Array<{ term: string }>;
    MF?: Array<{ term: string }>;
    CC?: Array<{ term: string }>;
  };
  pathway?: {
    reactome?: Array<{ name: string }> | { name: string };
    wikipathways?: Array<{ name: string }> | { name: string };
    kegg?: Array<{ name: string }> | { name: string };
  };
  MIM?: Array<{ MIM?: string; name?: string }> | { MIM?: string; name?: string };
  generif?: Array<{ pubmed?: number; text?: string }>;
  genomic_pos?:
    | { chr?: string; start?: number; end?: number }
    | Array<{ chr?: string; start?: number; end?: number }>;
  ensembl?: { gene?: string } | Array<{ gene?: string }>;
}

/**
 * Fetch gene data from MyGene.info API.
 *
 * @param geneSymbol - Gene symbol to query (e.g., "CTNND2", "BRCA1")
 * @returns GeneData object if successful, null on error or not found
 */
export async function fetchGeneData(geneSymbol: string): Promise<GeneData | null> {
  try {
    const queryUrl = `${MYGENE_BASE_URL}/query?q=symbol:${encodeURIComponent(
      geneSymbol
    )}&species=human&fields=entrezgene&size=1`;

    const queryResponse = await fetch(queryUrl);
    if (!queryResponse.ok) {
      console.warn(`MyGene.info query failed: ${queryResponse.statusText}`);
      return null;
    }

    const queryData: MyGeneQueryResponse = await queryResponse.json();

    if (!queryData.hits || queryData.hits.length === 0) {
      console.warn(`No data found for gene symbol: ${geneSymbol}`);
      return null;
    }

    const geneId = queryData.hits[0].entrezgene;
    if (!geneId) {
      console.warn(`No gene ID found for symbol: ${geneSymbol}`);
      return null;
    }

    const detailUrl = `${MYGENE_BASE_URL}/gene/${geneId}?fields=${MYGENE_FIELDS}`;
    const detailResponse = await fetch(detailUrl);
    if (!detailResponse.ok) {
      console.warn(`MyGene.info detail fetch failed: ${detailResponse.statusText}`);
      return null;
    }

    const detailData: MyGeneDetailResponse = await detailResponse.json();

    return parseMyGeneResponse(detailData, geneSymbol);
  } catch (error) {
    console.error(`Error fetching gene data for ${geneSymbol}:`, error);
    return null;
  }
}

/**
 * Parse MyGene.info API response into GeneData model.
 */
function parseMyGeneResponse(result: MyGeneDetailResponse, geneSymbol: string): GeneData {
  const goTerms: { BP?: string[]; MF?: string[]; CC?: string[] } = {};

  if (result.go) {
    if (result.go.BP) {
      goTerms.BP = result.go.BP.slice(0, 10).map((term) => term.term);
    }
    if (result.go.MF) {
      goTerms.MF = result.go.MF.slice(0, 10).map((term) => term.term);
    }
    if (result.go.CC) {
      goTerms.CC = result.go.CC.slice(0, 10).map((term) => term.term);
    }
  }

  const pathways: string[] = [];
  if (result.pathway) {
    for (const source of ['reactome', 'wikipathways', 'kegg'] as const) {
      const sourcePathways = result.pathway[source];
      if (sourcePathways) {
        if (Array.isArray(sourcePathways)) {
          for (const pw of sourcePathways) {
            if (pw.name) {
              pathways.push(`${pw.name} (${source.charAt(0).toUpperCase() + source.slice(1)})`);
            }
          }
        } else if (sourcePathways.name) {
          pathways.push(
            `${sourcePathways.name} (${source.charAt(0).toUpperCase() + source.slice(1)})`
          );
        }
      }
    }
  }

  const mimDiseases: string[] = [];
  if (result.MIM) {
    const mimData = Array.isArray(result.MIM) ? result.MIM : [result.MIM];
    for (const mim of mimData.slice(0, 5)) {
      if (mim.MIM || mim.name) {
        mimDiseases.push(mim.MIM && mim.name ? `MIM:${mim.MIM} - ${mim.name}` : mim.name || '');
      }
    }
  }

  const generif: string[] = [];
  if (result.generif && Array.isArray(result.generif)) {
    for (const rif of result.generif.slice(0, 5)) {
      if (rif.pubmed && rif.text) {
        generif.push(`PMID:${rif.pubmed}: ${rif.text.slice(0, 100)}...`);
      }
    }
  }

  let ensemblId: string | undefined;
  if (result.ensembl) {
    if (Array.isArray(result.ensembl) && result.ensembl.length > 0) {
      ensemblId = result.ensembl[0].gene;
    } else if (!Array.isArray(result.ensembl)) {
      ensemblId = result.ensembl.gene;
    }
  }

  let genomicLocation: string | undefined;
  if (result.genomic_pos) {
    const gpos = Array.isArray(result.genomic_pos) ? result.genomic_pos[0] : result.genomic_pos;
    if (gpos.chr && gpos.start && gpos.end) {
      genomicLocation = `chr${gpos.chr}:${gpos.start.toLocaleString()}-${gpos.end.toLocaleString()}`;
    }
  }

  return {
    geneSymbol: result.symbol || geneSymbol,
    geneName: result.name,
    summary: result.summary,
    goTerms: Object.keys(goTerms).length > 0 ? goTerms : undefined,
    pathways: pathways.length > 0 ? pathways : undefined,
    mimDiseases: mimDiseases.length > 0 ? mimDiseases : undefined,
    generif: generif.length > 0 ? generif : undefined,
    entrezgeneId: result.entrezgene,
    ensemblId,
    genomicLocation,
    source: 'mygene.info',
  };
}

/**
 * Format GeneData into a structured string for LLM context.
 */
export function formatGeneDataForLLM(geneData: GeneData): string {
  const lines: string[] = ['GENE DATABASE INFORMATION (MyGene.info):', ''];

  lines.push(`Gene: ${geneData.geneSymbol}`);
  if (geneData.geneName) {
    lines.push(`Full Name: ${geneData.geneName}`);
  }

  if (geneData.entrezgeneId) {
    lines.push(`NCBI Gene ID: ${geneData.entrezgeneId}`);
  }
  if (geneData.ensemblId) {
    lines.push(`Ensembl ID: ${geneData.ensemblId}`);
  }
  if (geneData.genomicLocation) {
    lines.push(`Genomic Location: ${geneData.genomicLocation}`);
  }

  lines.push('');

  if (geneData.summary) {
    lines.push('Summary:');
    lines.push(`  ${geneData.summary}`);
    lines.push('');
  }

  if (geneData.goTerms) {
    if (geneData.goTerms.BP && geneData.goTerms.BP.length > 0) {
      lines.push('Biological Processes (Gene Ontology):');
      for (const term of geneData.goTerms.BP.slice(0, 5)) {
        lines.push(`  - ${term}`);
      }
      lines.push('');
    }

    if (geneData.goTerms.MF && geneData.goTerms.MF.length > 0) {
      lines.push('Molecular Functions (Gene Ontology):');
      for (const term of geneData.goTerms.MF.slice(0, 5)) {
        lines.push(`  - ${term}`);
      }
      lines.push('');
    }

    if (geneData.goTerms.CC && geneData.goTerms.CC.length > 0) {
      lines.push('Cellular Components (Gene Ontology):');
      for (const term of geneData.goTerms.CC.slice(0, 5)) {
        lines.push(`  - ${term}`);
      }
      lines.push('');
    }
  }

  if (geneData.pathways && geneData.pathways.length > 0) {
    lines.push('Known Pathways:');
    for (const pathway of geneData.pathways.slice(0, 8)) {
      lines.push(`  - ${pathway}`);
    }
    lines.push('');
  }

  if (geneData.mimDiseases && geneData.mimDiseases.length > 0) {
    lines.push('Disease Associations (OMIM):');
    for (const disease of geneData.mimDiseases) {
      lines.push(`  - ${disease}`);
    }
    lines.push('');
  }

  if (geneData.generif && geneData.generif.length > 0) {
    lines.push('Recent Research (GeneRIF):');
    for (const rif of geneData.generif.slice(0, 3)) {
      lines.push(`  - ${rif}`);
    }
    lines.push('');
  }

  return lines.join('\n');
}
